flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferCompressor.java

```diff
@@ -0,0 +1,136 @@
+import org.apache.flink.runtime.io.compression.BlockCompressionFactory;
+import org.apache.flink.runtime.io.compression.BlockCompressor;
+
+import java.nio.ByteBuffer;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Compressor for {@link Buffer}.
+ */
+public class BufferCompressor {
+
+	/** The backing block compressor for data compression. */
+	private final BlockCompressor blockCompressor;
+
+	/** The intermediate heap buffer for the compressed data. */
+	private final byte[] heapBuffer;
+
+	public BufferCompressor(int bufferSize, String factoryName) {
+		checkArgument(bufferSize > 0);
+		// the size of this intermediate heap buffer will be gotten from the
+		// plugin configuration in the future, and currently, double size of
+		// the input buffer is enough for lz4-java compression library.
+		this.heapBuffer = new byte[2 * bufferSize];
+		this.blockCompressor = BlockCompressionFactory.createBlockCompressionFactory(factoryName).getCompressor();
+	}
+
+	/**
+	 * Compresses the given {@link Buffer} using {@link BlockCompressor}. The compressed data will be stored in the
+	 * internal heap buffer of this {@link BufferCompressor} and returned to the caller. The caller must guarantee
+	 * that the returned {@link Buffer} is freed when calling the method next time.
+	 *
+	 * <p>Notes that the compression will always start from offset 0 to the size of the input {@link Buffer}.
+	 */
+	public Buffer compressToInternalBuffer(Buffer buffer) {
+		int compressedLen;
+		if ((compressedLen = compress(buffer)) == 0) {
+			return buffer;
+		}
+
+		try {
+			if (compressedLen >= buffer.getSize()) {
+				return buffer;
+			}
+
+			// warp the internal heap buffer as Buffer
+			MemorySegment memorySegment = MemorySegmentFactory.wrap(heapBuffer);
+			NetworkBuffer compressedBuffer = new NetworkBuffer(memorySegment, FreeingBufferRecycler.INSTANCE);
+			compressedBuffer.setSize(compressedLen);
+			compressedBuffer.setCompressed(true);
+
+			return compressedBuffer;
+		} catch (Throwable throwable) {
+			return buffer;
+		}
+	}
+
+	/**
+	 * The difference between this method and {@link #compressToInternalBuffer(Buffer)} is that this method will
+	 * copy the compressed data back to the input {@link Buffer} starting from offset 0.
+	 *
+	 * <p>The caller must guarantee that the input {@link Buffer} is writable and there's enough space left.
+	 */
+	public Buffer compressInPlace(Buffer buffer) {
+		int compressedLen;
+		if ((compressedLen = compress(buffer)) == 0) {
+			return buffer;
+		}
+
+		try {
+			if (compressedLen >= buffer.getSize()) {
+				return buffer;
+			}
+
+			// copy the compressed data back
+			MemorySegment segment = buffer.getMemorySegment();
+			segment.put(buffer.getMemorySegmentOffset(), heapBuffer, 0, compressedLen);
+			Buffer compressedBuffer = buffer.readOnlySlice(0, compressedLen);
+			compressedBuffer.setCompressed(true);
+
+			return compressedBuffer;
```

Comment:
`return new ReadOnlySlicedNetworkBuffer(buffer, 0, compressedLen, true)` for replacing above three clauses.